packages/go_router/lib/src/delegate.dart

```diff
@@ -67,15 +75,27 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
 
     _matches.push(newPageKeyMatch);
     notifyListeners();
+    return completer.future;
   }
 
   /// Returns `true` if there is more than 1 page on the stack.
   bool canPop() {
     return _matches.canPop();
   }
 
-  /// Pop the top page off the GoRouter's page stack.
-  void pop() {
+  /// Pop the top page off the GoRouter's page stack and complete a promise if
+  /// there is one.
+  void pop<T extends Object?>([T? value]) {
+    final RouteMatch last = _matches.last;
+
+    // If there is a promise for this page, complete it.
+    final Completer<T>? completer = _completers[last.fullpath] as Completer<T>?;
+    if (completer != null) {
+      completer.complete(value);
+    }
+
+    // Remove promise from completers map.
+    _completers.remove(last.fullpath);
```

Comment:
I think the completer should be removed from the list as well when users call `void replace(RouteMatch match)`. Because the last RouteMatch was removed.